.github/workflows/coverage.yaml

```diff
@@ -0,0 +1,131 @@
+
+      - name: Restore cached keys
+        uses: actions/cache/restore@v3
+        with:
+          path: keys
+          key: keys-folder
+
+      - name: Install LXD-UI dependencies
+        run: |
+          set -x
+          sudo chmod 0777 ../lxd-ui
+          dotrun install
+
+      - name: Run LXD-UI
+        env:
+          ENVIRONMENT: devel
+          PORT: 8407
+          LXD_UI_BACKEND_IP: 172.17.0.1
+        run: |
+          dotrun &
+          curl --head --fail --retry-delay 2 --retry 100 --retry-connrefused --insecure https://localhost:8407
+
+      - name: Set keys permissions
+        run: |
+          set -x
+          sudo chmod -R 0666 keys
+          sudo chmod 0777 keys
+          
+      - name: Save keys
+        uses: actions/cache/save@v3
+        with:
+          path: keys
+          key: keys-folder
+
+      - name: Install LXD
+        uses: canonical/setup-lxd@v0.1.1
+        with:
+          channel: "latest/edge"
+
+      - name: Setup LXD
+        shell: bash
+        run: |
+          set -x
+          sudo lxc config set core.https_address "[::]:8443"
+          sudo lxc config trust add keys/lxd-ui.crt
+          sudo lxc config set cluster.https_address "127.0.0.1"
+          sudo lxc cluster enable local
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 18
+
+      - name: Install Playwright Browser
+        run: npx playwright install --with-deps chromium
+
+      - name: Run tests with coverage
+        run: yarn test-e2e-coverage
+
+      - name: Upload coverage report
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: lxd-coverage
+          path: coverage/playwright-report
+          retention-days: 1
+
+  publish-coverage-report:
+      name: publish-coverage-report
+      runs-on: ubuntu-latest
+      needs: e2e-coverage
+      continue-on-error: true
+      steps:
+        - uses: actions/checkout@v4
+          with:
+            ref: gh-pages
+            token: ${{ secrets.GITHUB_TOKEN }}
+        - name: Cleanup coverage directory
+          run: |
+            rm -rf coverage
+            mkdir coverage
+        - name: Download coverage report artifact
+          uses: actions/download-artifact@v4
+          with:
+            name: lxd-coverage
+            path: coverage
+        # user git configs are needed for git commands to work
+        # actual authentication is done using secrets.GITHUB_TOKEN with write permission
+        - name: Set Git User
+          run: |
+            git config --global user.email "github-action@example.com"
+            git config --global user.name "GitHub Action"
+        - name: Push coverage Report
+          timeout-minutes: 3
+          run: |
+            git add .
+            git commit -m "workflow: update coverage report"
+            
+            # In case of another action job pushing to gh-pages while we are rebasing for the current job
```

Comment:
not sure if this check is necessary or is the right approach in this scenario. For normal e2e reports, they are stored in folders on `gh-pages` branch separated by their origin PR number as well as workflow run id. In that case, we don't want to completely stop a workflow run just because another workflow run is busy with the `gh-pages` branch. In the case of the coverage report, I think what we need to ensure is that a later workflow run does not finish earlier than prior workflow runs (e.g. two PRs closed in short succession). Not sure how this can be done though since the push event trigger completely separates the workflow run context...

Reply:
This solves the problem of other actions pushing to gh-pages, that could be another PR test report.

There might be two merges in timely proximitry. If the 1st coverage is pushed after the 2nd coverage, we'd get a report that is slightly outdated. Though not a big issue I think as the next merge will trigger another new coverage report.

Reply:
Okay yeah I think let's leave it as is, slightly outdated coverage reports in this case wouldn't be a major issue in any case :+1: